pkg/controller/virtualmachine/virtualmachine.go

```diff
@@ -0,0 +1,475 @@
+	return nil
+}
+
+// OnVMIChange attempts to reconcile devices allocated to launcher pod associated with a running VMI
+// and annotate the VM object with device annotations
+func (h *Handler) OnVMIChange(name string, vmi *kubevirtv1.VirtualMachineInstance) (*kubevirtv1.VirtualMachineInstance, error) {
+	if vmi == nil || vmi.DeletionTimestamp != nil {
+		logrus.WithFields(logrus.Fields{
+			"name": name,
+		}).Debug("skipping object, either does not exist or marked for deletion")
+		return vmi, nil
+	}
+
+	logrus.WithFields(logrus.Fields{
+		"name":      vmi.Name,
+		"namespace": vmi.Namespace,
+	}).Debug("reconcilling vmi device allocation")
+
+	if vmi.Status.Phase != kubevirtv1.Running {
+		logrus.WithFields(logrus.Fields{
+			"name":      vmi.Name,
+			"namespace": vmi.Namespace,
+		}).Debug("skipping vmi as it is not running")
+		return vmi, nil
+	}
+
+	if len(vmi.Spec.Domain.Devices.HostDevices) == 0 && len(vmi.Spec.Domain.Devices.GPUs) == 0 {
+		logrus.WithFields(logrus.Fields{
+			"name":      vmi.Name,
+			"namespace": vmi.Namespace,
+		}).Debug("skipping vmi as it does not request any host devices or gpus")
+		return vmi, h.checkAndClearDeviceAllocation(vmi)
+	}
+
+	if vmi.Status.NodeName != h.nodeName {
+		logrus.WithFields(logrus.Fields{
+			"name":      vmi.Name,
+			"namespace": vmi.Namespace,
+			"hostname":  vmi.Status.NodeName,
+		}).Debug("skipping vmi as it is not scheduled on current node")
+		return vmi, nil
+	}
+	return vmi, h.trackDevices(vmi)
+}
+
+// trackDevices reconciles GPU and HostDevices info
+func (h *Handler) trackDevices(vmi *kubevirtv1.VirtualMachineInstance) error {
+	envMap, err := h.generatePodEnvMap(vmi)
+	if err != nil {
+		return err
+	}
+	return h.reconcileDeviceAllocationDetails(vmi, envMap)
+}
+
+// reconcileDeviceAllocationDetails will reconcile envMap into device allocation annotation on vmi
+// has been split into its own method to simplify testing
+func (h *Handler) reconcileDeviceAllocationDetails(vmi *kubevirtv1.VirtualMachineInstance, envMap map[string]string) error {
+	var pciDeviceMap, vGPUMap map[string]string
+	selector := map[string]string{
+		"nodename": vmi.Status.NodeName,
+	}
+	if len(vmi.Spec.Domain.Devices.HostDevices) > 0 {
+		deviceList, err := h.pciDeviceCache.List(labels.SelectorFromSet(selector))
+		if err != nil {
+			return fmt.Errorf("error listing pcidevices from cache: %v", err)
+		}
+		pciDeviceMap = buildPCIDeviceMap(deviceList)
+	}
+	if len(vmi.Spec.Domain.Devices.GPUs) > 0 {
+		deviceList, err := h.vgpuCache.List(labels.SelectorFromSet(selector))
+		if err != nil {
+			return fmt.Errorf("error listing vgpudevices from cache: %v", err)
+		}
+		vGPUMap = buildVGPUMap(deviceList)
+	}
+	// map to hold device details
+	hostDeviceMap := make(map[string][]string)
+	gpuMap := make(map[string][]string)
+	for _, device := range vmi.Spec.Domain.Devices.HostDevices {
+		val, ok := envMap[util.ResourceNameToEnvVar(deviceplugins.PCIResourcePrefix, device.DeviceName)]
```

Comment:
Just confirm that. Since the usbdevice is not like pcidevice and vGPU device are randomly picked up by KubeVirt, so we can skip collecting envMap from `USBResourcePrefix`. If yes, do we need to add some comments about why we don't need to collect from `USBResourcePrefix`?

Reply:
yes for usb devices a new plugin is created for each device so it will not suffer from this issue